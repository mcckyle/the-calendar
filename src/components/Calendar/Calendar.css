/* Filename: Calendar.css */

/* ================================
    Calendar Wrapper Styling.
=================================== */
.calendar
{
    width: 100%;
    margin: 0 auto;

    display: flex;
    flex-direction: column;
    gap: 1.25rem;

    background: var(--secondary-bg);
    border-radius: var(--border-radius);
    border: 1px solid #e5e73b;
    padding: 1.5rem 2rem;
    box-sizing: border-box;

    animation: fadeIn 0.5s ease-out forwards;
}

/* =============================
    Calendar Header Styling.
================================ */
.calendar-header
{
    display: flex;
    align-items: center;
    justify-content: center;

    padding: 0.75rem 1rem;
    border-bottom: 1px solid rgba(0, 0, 0, 0.05);

    position: sticky;
    top: 0;
    z-index: 10;

    background: rgba(27, 31  59, 0.92);
    backdrop-filter: blur(12px);
    border-radius: var(--border-radius) var(--border-radius) 0 0;

    font-family: var(--title-font);
    font-size: 1.1rem;
    font-weight: 600;
    color: #fff;
    letter-spacing: 0.3px;
}

/* ===================================
   Calendar Status Messages Styles.
====================================== */
.calendar-status
{
    text-align: center;
    font-size: 0.9rem;
    padding: 0.6rem 1rem;
    border-radius: var(--border-radius);
    background: rgba(0, 0, 0, 0.03);
    color: var(--text-color);
    max-width: 400px;
    margin: 0 auto;
}

.calendar-status.error
{
    background: rgba(255, 0, 0, 0.08);
    color: #b00020;
    font-weight: 500;
}

/* ========================================
    Calendar Body (7 Columns) Styling.
 ========================================== */
.calendar-body
{
    width: 100%;
    overflow-x: auto;
    -webkit-overflow-scrolling: touch;
    padding-bottom: 0.5rem;
}

/* =======================================
    Calendar Week Grid Layout Styling.
========================================== */
.week-view
{
    display: grid;
    grid-template-columns: repeat(7, minmax(160px, 1fr));
    gap: 0.75rem;

    min-width: calc(7 * 160px + 6 * 0.75rem);
}

.week-view > *
{
    transition: transform 0.2s ease, box-shadow 0.2s ease;
}

.week-view > *:hover
{
    transform: translateY(-2px);
    box-shadow: var(--shadow-sm);
}

/* =========================
    Animations Styling.
============================ */
@keyframes fadeIn
{
    from
    {
        opacity: 0;
        transform: translateY(10px);
    }
    to
    {
        opacity: 1;
        transform: translateY(0);
    }
}
