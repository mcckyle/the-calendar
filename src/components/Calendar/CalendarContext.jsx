//Filename: CalendarContext.jsx
//Author: Kyle McColgan
//Date: 04 September 2025
//Description: This file contains the Calendar context component for the Saint Louis React calendar project.

import React, { createContext, useContext, useState } from 'react';

export const CalendarContext = createContext();
export const useCalendarContext = () => useContext(CalendarContext);

const getStartOfWeek = (date) => {
  const newDate = new Date(date);
  const day = newDate.getDay(); //Sunday == 0, Monday == 1, etc.
  newDate.setDate(newDate.getDate() - day); //Backup to Sunday.
  return newDate;
};

export const CalendarProvider = ({ children }) => {
  const [currentDate, setCurrentDate] = useState(getStartOfWeek(new Date()));
  const [selectedDate, setSelectedDate] = useState(null);

  const changeMonth = (offset) => {
    const newDate = new Date(currentDate);
    newDate.setMonth(currentDate.getMonth() + offset);
    setCurrentDate(getStartOfWeek(newDate));
  };

  const changeWeek = (offset) => {
    const newDate = new Date(currentDate);
    newDate.setDate(currentDate.getDate() + offset * 7);
    setCurrentDate(getStartOfWeek(newDate));
  };

  const selectDate = (date) => {
    const parsedDate = new Date(date);
    if (!isNaN(parsedDate)) {
      setSelectedDate(parsedDate);
    }
  };

  return (
    <CalendarContext.Provider
      value={{
        currentDate,
        selectedDate, // Add selectedDate to the context value.
        changeMonth,
        changeWeek,
        selectDate, // Add selectDate function.
      }}
    >
      {children}
    </CalendarContext.Provider>
  );
};
