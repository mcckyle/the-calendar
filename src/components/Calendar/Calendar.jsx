//Filename: Calendar.jsx
//Author: Kyle McColgan
//Date: 04 September 2025
//Description: This file contains the parent component for the Saint Louis calendar project.

import React, { useState, useEffect } from 'react';
import { useCalendarContext } from './CalendarContext';
import { useEvents } from "../../hooks/useEvents";
import DaysOfWeek from "../DaysOfWeek/DaysOfWeek.jsx";
import events from '../../data/events.json';
import EventPanel from '../EventPanel/EventPanel.jsx';
import WeekNavigation from '../WeekNavigation/WeekNavigation.jsx';
import WeekDayColumn from '../WeekDayColumn/WeekDayColumn.jsx';

import {
	convertTo12HourFormat,
	normalizeEvents,
	groupEventsByHour,
} from "../../utils/eventUtils";

import './Calendar.css';

const Calendar = ({ hours }) => {
  const [selectedEvent, setSelectedEvent] = useState(null);
  const [showEventPanel, setShowEventPanel] = useState(false);
  const [weekDays, setWeekDays] = useState([]);
  const { currentDate } = useCalendarContext();
  const [normalizedEvents, setNormalizedEvents] = useState([]);

  const apiUrl = "https://calendar-backend-xxa6.onrender.com";

  //Compute the start + end of the current week from CalendarContext.jsx.
  const weekStart = currentDate.toISOString().split("T")[0];
  const weekEnd = new Date(currentDate);
  weekEnd.setDate(currentDate.getDate() + 6); // Sunday + 6 == Saturday.
  const weekEndISO = weekEnd.toISOString().split("T")[0];

  //Fetch the events...
  const { events, loading, error } = useEvents(apiUrl, weekStart, weekEndISO);

  useEffect(() => {
    if( (events) && (events.length > 0) )
    {
      const normalized = normalizeEvents(events);
      console.log("Normalized Events:", normalized); //For debugging purposes...
      setNormalizedEvents(normalized);
    }
    else
    {
      setNormalizedEvents([]); //Clear if no events found...
    }
  }, [events]);

  //Generate the seven days of the week based on CalendarContext.jsx.
  useEffect(() => {
    const generateWeekDays = () => {
      const weekDaysArray = Array.from({ length: 7 }, (_, i) => {
        const day = new Date(currentDate);
        day.setDate(currentDate.getDate() + i);
        return day;
      });
      setWeekDays(weekDaysArray);
      console.log("Week Dates:", weekDaysArray); // Log to verify correct week days.
    };

    generateWeekDays();
  }, [currentDate]);

  const handleEventClick = (event) => {
    setSelectedEvent(event);
    setShowEventPanel(true); // Open the event panel.
  };

  const closeEventPanel = () => {
    setShowEventPanel(false);
    setSelectedEvent(null);
  };

  return (
    <div className="calendar-container">
      <div className="calendar-header">
        <WeekNavigation />
      </div>

      {loading && <p>Loading events...</p>}
      {error && <p style={{ color: "red" }}>Error: {error}</p>}
      
      <div className="calendar-body">
        <DaysOfWeek weekDays={weekDays} />
        <div className="week-view">
          {weekDays.map((day) => (
            <WeekDayColumn
              key={day.toDateString()}
              day={day}
              groupedEvents={groupEventsByHour(day, normalizedEvents)}
              onEventClick={handleEventClick}
              convertTo12HourFormat={convertTo12HourFormat}
            />
          ))}
        </div>
      </div>

      {/* Conditionally render the EventPanel */}
      {showEventPanel && (
        <EventPanel
          selectedEvent={selectedEvent}
          onClose={closeEventPanel}
        />
      )}
    </div>
  );
};

export default Calendar;
