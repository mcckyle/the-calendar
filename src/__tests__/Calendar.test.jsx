//Filename: Calendar.test.jsx
//Author: Kyle McColgan
//Date: 05 September 2025
//Description: This file contains unit tests for the Calendar.jsx component.

import React from 'react';
import '@testing-library/jest-dom';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import Calendar from '../components/Calendar/Calendar.jsx';
import { CalendarProvider } from '../components/Calendar/CalendarContext';

//Mock WeekDayColumn.jsx.
jest.mock('../components/WeekDayColumn/WeekDayColumn.jsx', () => ({ day }) => (
  <div data-testid="week-day-column">
    <span>{day.toDateString()}</span>
  </div>
));

//Mock EventPanel.jsx.
jest.mock('../components/EventPanel/EventPanel.jsx', () => ({ selectedEvent, onClose }) => (
  <div data-testid="event-panel">
    <span>{selectedEvent ? selectedEvent.title : 'No Event'}</span>
    <button onClick={onClose}>Close</button>
  </div>
));

describe('Calendar Component', () => {
  const renderCalendar = () => {
    render(
      <CalendarProvider>
        <Calendar />
      </CalendarProvider>
    );
  };

  //Test #1: Header renders correctly.
  test('renders calendar header with navigation buttons and month/year.', () => {
    renderCalendar();

    // Check for navigation buttons and month/year display.
    expect(screen.getByRole('button', { name: /previous week/i })).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /next week/i })).toBeInTheDocument();
    expect(screen.getByText(/January|February|March|April|May|June|July|August|September|October|November|December/)).toBeInTheDocument();
  });

  //Test #2: Week view renders seven (7) weekday columns.
  test('renders week view with seven WeekDayColumn components.', () => {
    renderCalendar();

    // Check that seven WeekDayColumn components are rendered.
    expect(screen.getAllByTestId('week-day-column').length).toBe(7);
  });

  //Test #3: Clicking "Previous" keeps seven (7) columns but changes dates.
  test('navigates to the previous week when the "Previous" button is clicked.', () => {
    renderCalendar();

    // Click on the "Next week" button using the accessible name.
    fireEvent.click(screen.getByRole('button', { name: /previous week/i }));

    // Verify that the week updates.
    expect(screen.getAllByTestId('week-day-column').length).toBe(7); // Week view remains consistent.
  });

  //Test #4: Clicking "Next" changes week dates.
  test('navigates to the next week when the "Next" button is clicked.', () => {
    renderCalendar();

    // Click on the "Next week" button using the accessible name.
    fireEvent.click(screen.getByRole('button', { name: /next week/i }));

    // Verify that the week updates.
    expect(screen.getAllByTestId('week-day-column').length).toBe(7); // Week view remains consistent.
  });

   //Test #5: Month text updates when navigating.
  test('updates the displayed month/year when navigating weeks.', () => {
    renderCalendar();
    const monthBefore = screen.getByText(/January|February|March|April|May|June|July|August|September|October|November|December/).textContent;

    for(let i = 0; i < 5; i ++ )
    {
        // Click on the "Next week" button using the accessible name.
        fireEvent.click(screen.getByRole('button', { name: /next week/i }));
    }

    const monthAfter = screen.getByText(/January|February|March|April|May|June|July|August|September|October|November|December/).textContent;

    expect(monthBefore).not.toEqual(monthAfter);
  });

  //Test #6: EventPanel does not render initially.
  test('does not render EventPanel initially.', () => {
    renderCalendar();
    expect(screen.queryByTestId('event-panel')).toBeNull();
  });

  //Test #7: EventPanel renders when showEventPanel is true.
  test('renders EventPanel when an event is clicked on.', () => {
    renderCalendar();

    // Click on the "Next week" button using the accessible name.
    fireEvent.click(screen.getByRole('button', { name: /next week/i }));

    //Force showing EventPanel.
    const { rerender } = render(
      <CalendarProvider>
        <Calendar />
      </CalendarProvider>
    );

    rerender(
      <CalendarProvider>
        <Calendar />
      </CalendarProvider>
    );

    expect(screen.queryByTestId('event-panel')).toBeNull();
  });

  //Test #8: Weekday column displays correct date format.
  test('displays weekday columns with readable date strings.', () => {
    renderCalendar();
    const columns = screen.getAllByTestId('week-day-column');
    columns.forEach(column => {
      expect(column.textContent).toMatch(/\w+\s\d{1,2}\s\d{4}/);
    });
  });

  //Test #9: Snapshot test for UI consistency.
  test('matches snapshot for default render.', () => {
    const { asFragment } = render(
      <CalendarProvider>
        <Calendar />
      </CalendarProvider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  //Test #10: Accessibility - buttons have accessible roles.
  test('navigation buttons are accessible', () => {
    renderCalendar();
    expect(screen.getByRole('button', { name: /Previous/i})).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /Next/i})).toBeInTheDocument();
  });
});
