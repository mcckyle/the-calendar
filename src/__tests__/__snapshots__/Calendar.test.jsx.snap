// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar Component matches snapshot for default render. 1`] = `
<DocumentFragment>
  <section
    class="calendar"
  >
    <header
      class="calendar-header"
    >
      <nav
        aria-label="Week navigation"
        class="week-navigation"
      >
        <button
          aria-label="Previous week"
          class="nav-button"
        >
          ‹
        </button>
        <time
          aria-live="polite"
          class="month-year"
        >
          September 2025
        </time>
        <button
          aria-label="Next week"
          class="nav-button"
        >
          ›
        </button>
      </nav>
    </header>
    <p
      class="calendar-status error"
    >
      Error: fetch is not defined
    </p>
    <div
      class="calendar-body"
    >
      <div
        class="days-of-week-container"
        role="row"
      >
        <div
          aria-label="Sunday, September 14"
          class="days-of-week-item "
          role="columnheader"
        >
          <span
            class="dow-label"
          >
            Sun
          </span>
          <span
            class="dow-date"
          >
            Sep 14
          </span>
        </div>
        <div
          aria-label="Monday, September 15"
          class="days-of-week-item today"
          role="columnheader"
        >
          <span
            class="dow-label"
          >
            Mon
          </span>
          <span
            class="dow-date"
          >
            Sep 15
          </span>
        </div>
        <div
          aria-label="Tuesday, September 16"
          class="days-of-week-item "
          role="columnheader"
        >
          <span
            class="dow-label"
          >
            Tue
          </span>
          <span
            class="dow-date"
          >
            Sep 16
          </span>
        </div>
        <div
          aria-label="Wednesday, September 17"
          class="days-of-week-item "
          role="columnheader"
        >
          <span
            class="dow-label"
          >
            Wed
          </span>
          <span
            class="dow-date"
          >
            Sep 17
          </span>
        </div>
        <div
          aria-label="Thursday, September 18"
          class="days-of-week-item "
          role="columnheader"
        >
          <span
            class="dow-label"
          >
            Thu
          </span>
          <span
            class="dow-date"
          >
            Sep 18
          </span>
        </div>
        <div
          aria-label="Friday, September 19"
          class="days-of-week-item "
          role="columnheader"
        >
          <span
            class="dow-label"
          >
            Fri
          </span>
          <span
            class="dow-date"
          >
            Sep 19
          </span>
        </div>
        <div
          aria-label="Saturday, September 20"
          class="days-of-week-item "
          role="columnheader"
        >
          <span
            class="dow-label"
          >
            Sat
          </span>
          <span
            class="dow-date"
          >
            Sep 20
          </span>
        </div>
      </div>
      <div
        class="week-view"
      >
        <div
          data-testid="week-day-column"
        >
          <span>
            Sun Sep 14 2025
          </span>
        </div>
        <div
          data-testid="week-day-column"
        >
          <span>
            Mon Sep 15 2025
          </span>
        </div>
        <div
          data-testid="week-day-column"
        >
          <span>
            Tue Sep 16 2025
          </span>
        </div>
        <div
          data-testid="week-day-column"
        >
          <span>
            Wed Sep 17 2025
          </span>
        </div>
        <div
          data-testid="week-day-column"
        >
          <span>
            Thu Sep 18 2025
          </span>
        </div>
        <div
          data-testid="week-day-column"
        >
          <span>
            Fri Sep 19 2025
          </span>
        </div>
        <div
          data-testid="week-day-column"
        >
          <span>
            Sat Sep 20 2025
          </span>
        </div>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
